File: Demo/Demo/ViewController.m
//
//  ViewController.m
//  Demo
//
//  Created by drif on 11/27/16.
//  Copyright © 2016 Payfort (https://start.payfort.com). All rights reserved.
//

@import StartSDK;

#import "ViewController.h"

@interface ViewController ()

@property (nonatomic, weak) IBOutlet UILabel *amountLabel;
@property (nonatomic, weak) IBOutlet UITextField *numberTextField;
@property (nonatomic, weak) IBOutlet UITextField *monthTextField;
@property (nonatomic, weak) IBOutlet UITextField *yearTextField;
@property (nonatomic, weak) IBOutlet UITextField *cvcTextField;
@property (nonatomic, weak) IBOutlet UITextField *cardholderTextField;
@property (nonatomic, weak) IBOutlet UILabel *errorsLabel;
@property (nonatomic, weak) IBOutlet UIActivityIndicatorView *activityIndicatorView;
@property (nonatomic, strong) IBOutletCollection(UITextField) NSArray *textFields;

@end

@implementation ViewController {
    Start *_start;
    NSInteger _amount;
    NSString *_currency;
}

#pragma mark - Private methods

- (NSString *)amountString {
    NSNumberFormatter *formatter = [[NSNumberFormatter alloc] init];
    formatter.numberStyle = NSNumberFormatterCurrencyStyle;
    formatter.currencyCode = _currency;
    return [formatter stringFromNumber:@(_amount / 100.0f)];
}

- (StartCard *)card {
    NSError *error;
    NSMutableString *errorText;

    NSString *expirationYearString = [@"20" stringByAppendingString:(NSString *_Nonnull) self.yearTextField.text];

    StartCard *card = [StartCard cardWithCardholder:(NSString *_Nonnull) self.cardholderTextField.text
                                             number:(NSString *_Nonnull) self.numberTextField.text
                                                cvc:(NSString *_Nonnull) self.cvcTextField.text
                                    expirationMonth:self.monthTextField.text.integerValue
                                     expirationYear:expirationYearString.integerValue
                                              error:&error];

    if (error) {
        NSDictionary *fields = @{
                StartCardValueCardholder: self.cardholderTextField.placeholder,
                StartCardValueNumber: self.numberTextField.placeholder,
                StartCardValueCVC: self.cvcTextField.placeholder,
                StartCardValueExpirationYear: self.yearTextField.placeholder,
                StartCardValueExpirationMonth: self.monthTextField.placeholder
        };

        errorText = [NSMutableString stringWithString:@"The following fields are invalid:"];
        for (NSString *value in error.userInfo[StartCardErrorKeyValues]) {
            [errorText appendFormat:@"\n%@", fields[value]];
        }
    }

    self.errorsLabel.text = errorText;

    return card;
}

- (void)pay {
    StartCard *card = self.card;
    if (!card) {
        return;
    }

    [self resignFirstResponder];

    [self.activityIndicatorView startAnimating];
    [_start createTokenForCard:card amount:_amount currency:_currency successBlock:^(id <StartToken> token) {
        [self showAlertWithText:[NSString stringWithFormat:@"Token ID received from server:\n%@", token.tokenId]];
        [self.activityIndicatorView stopAnimating];
    } errorBlock:^(NSError *error) {
        [self showAlertWithText:[NSString stringWithFormat:@"Error occurred:\n%@", error]];
        [self.activityIndicatorView stopAnimating];
    } cancelBlock:^{
        [self showAlertWithText:@"Cancelled"];
        [self.activityIndicatorView stopAnimating];
    }];
}

- (void)showAlertWithText:(NSString *)text {
    UIAlertController *controller = [UIAlertController alertControllerWithTitle:nil message:text preferredStyle:UIAlertControllerStyleAlert];
    [controller addAction:[UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:^(UIAlertAction *action) {
        [self dismissViewControllerAnimated:YES completion:nil];
    }]];
    [self presentViewController:controller animated:YES completion:nil];
}

- (IBAction)onPay:(UIButton *)sender {
    [self pay];
}

- (IBAction)onTap:(UITapGestureRecognizer *)sender {
    [self resignFirstResponder];
}

#pragma mark - UIResponder methods

- (BOOL)resignFirstResponder {
    for (UITextField *textField in self.textFields) {
        [textField resignFirstResponder];
    }
    return [super resignFirstResponder];
}

#pragma mark - UIViewController methods

- (void)viewDidLoad {
    [super viewDidLoad];

    _start = [Start startWithAPIKey:@"test_open_k_46dd87e36d3a5949aa68"];
    _amount = 100;
    _currency = @"USD";
    
    self.amountLabel.text = [@"Pay " stringByAppendingString:self.amountString];

#if DEBUG
    self.numberTextField.text = @"4242424242424242";
    self.monthTextField.text = @"11";
    self.yearTextField.text = @"16";
    self.cvcTextField.text = @"123";
    self.cardholderTextField.text = @"John Smith";
#endif
}

@end

File: DemoSwift/DemoSwift/ViewController.swift
//
//  ViewController.swift
//  DemoSwift
//
//  Created by drif on 11/30/16.
//  Copyright © 2016 Payfort (https://start.payfort.com). All rights reserved.
//

import UIKit
import StartSDK

class ViewController: UIViewController {
    
    @IBOutlet weak var numberTextField: UITextField?
    @IBOutlet weak var monthTextField: UITextField?
    @IBOutlet weak var yearTextField: UITextField?
    @IBOutlet weak var cvcTextField: UITextField?
    @IBOutlet weak var cardholderTextField: UITextField?
    @IBOutlet weak var errorsLabel: UILabel?
    @IBOutlet weak var activityIndicatorView: UIActivityIndicatorView?
    @IBOutlet var textFields: [UITextField]?
    
    @IBOutlet weak var amountLabel: UILabel? {
        didSet {
            amountLabel?.text = "Pay \(amountString)"
        }
    }
    
    @IBAction func onPay(_ sender: UIButton) {
        pay()
    }
    @IBAction func onTap(_ sender: UITapGestureRecognizer) {
        let _ = resignFirstResponder()
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        #if DEBUG
            numberTextField?.text = "4242424242424242";
            monthTextField?.text = "11";
            yearTextField?.text = "16";
            cvcTextField?.text = "123";
            cardholderTextField?.text = "John Smith";
        #endif
    }
    
    override func resignFirstResponder() -> Bool {
        textFields?.forEach { $0.resignFirstResponder() }
        return super.resignFirstResponder()
    }

    let amount = 100
    let currency = "USD"
    let start = Start(apiKey: "test_open_k_46dd87e36d3a5949aa68")

    var amountString: String {
        let amountNumber = NSNumber(floatLiteral: Double(amount) / 100)
        
        let formatter = NumberFormatter()
        formatter.numberStyle = .currency
        formatter.currencyCode = currency
        return formatter.string(from: amountNumber) ?? ""
    }

    var card: StartCard? {
        var card: StartCard?
        var errorString: String?
        
        do {
            let cardholder = cardholderTextField?.text ?? ""
            let number = numberTextField?.text ?? ""
            let cvc = cvcTextField?.text ?? ""
            let month = Int(monthTextField?.text ?? "") ?? 0
            let year = Int("20\(yearTextField?.text ?? "")") ?? 0

            card = try StartCard(cardholder: cardholder, number: number, cvc: cvc, expirationMonth: month, expirationYear: year)
        }
        catch let error as NSError {
            if let errors = error.userInfo[StartCardErrorKeyValues] as? [String] {
                errorString = "The following fields are invalid:"
                errors.forEach { errorString = "\(errorString ?? "")\n\($0)" }
            }
            else {
                errorString = "Unknown error occured"
            }
        }
        catch {
            errorString = "Unknown error occured"
        }
        
        errorsLabel?.text = errorString
        
        return card
    }

    func pay() {
        guard let card = card else {
            return
        }

        let _ = resignFirstResponder()

        activityIndicatorView?.startAnimating()
        start.createToken(for: card, amount: amount, currency: currency, successBlock: { token in
            self.showAlert(with: "Token ID received from server:\n\(token.tokenId)")
            self.activityIndicatorView?.stopAnimating()
        }, errorBlock: { error in
            self.showAlert(with: "Error occured:\n\(error)")
            self.activityIndicatorView?.stopAnimating()
        }, cancel: {
            self.showAlert(with: "Cancelled")
            self.activityIndicatorView?.stopAnimating()
        })
    }
    
    func showAlert(with text: String) {
        let action = UIAlertAction(title: "OK", style: .default) { _ in
            self.dismiss(animated: true, completion: nil)
        }
        let controller = UIAlertController(title: nil, message: text, preferredStyle: .alert)
        controller.addAction(action)
        present(controller, animated: true, completion: nil)
    }
}
